src/_pytest/pathlib.py

```diff
@@ -607,11 +607,20 @@ def _is_same(f1: str, f2: str) -> bool:
 
 def module_name_from_path(path: Path, root: Path) -> str:
     """
-    Return a dotted module name based on the given path, anchored on root.
+    Return a dotted module name based on the given path,
+    anchored on root or the most likely entry in `sys.path`.
 
     For example: path="projects/src/tests/test_foo.py" and root="/projects", the
     resulting module name will be "src.tests.test_foo".
     """
+    candidates = (
+        _module_name_from_path(path, dir)
+        for dir in itertools.chain([root], map(Path, sys.path))
+    )
+    return ".".join(min(candidates, key=len))  # type: ignore[arg-type]
+
+
+def _module_name_from_path(path: Path, root: Path) -> "tuple[str, ...]":
```

Comment:
It's a bit confusing `module_name_from_path` and `_module_name_from_path`, should use some different name.

Reply:
makes sense!